### PR TITLE
br: refactoring code - BR may panic when error is not able to convert into a terror

### LINKED ISSUE #32293: br: batch ddl restore may panic when BatchCreateTableWithInfo failure with some special error
### 1. Minimal reproduce step 
1. Inject a simple error out of mysql error code
2. do a batch ddl restore

### 2. What did you expect to see? 
Restore will failure without panic
### 3. What did you see instead
Restore failure with panic

### PR DESCRIPTION
**Pull request by fengou1**
### What problem does this PR solve?

Issue Number: close #32293 

Problem Summary:

### What is changed and how it works?
break a cascade if condition into a function, so that we may take error type then compare the error code, this make BR robust
### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: br/pkg/restore/client.go [hunk 1]
```diff
@@ -617,6 +617,11 @@ func (rc *Client) createTablesInWorkerPool(ctx context.Context, dom *domain.Doma
 		workers.ApplyWithIDInErrorGroup(eg, func(id uint64) error {
 			db := dbPool[id%uint64(len(dbPool))]
 			cts, err := rc.createTables(ectx, db, dom, tableSlice, newTS) // ddl job for [lastSent:i)
+			failpoint.Inject("restore-createtables-error", func(val failpoint.Value) {
+				if val.(bool) {
+					err = errors.Trace(fmt.Errorf("sample error without extra message"))
```
Comment (kennytm) [inline br/pkg/restore/client.go:622]:
```suggestion
					err = errors.New("sample error without extra message")
```

File: br/pkg/utils/retry.go [hunk 2]
```diff
@@ -138,3 +139,13 @@ func isSingleRetryableError(err error) bool {
 		}
 	}
 }
+
+func FallBack2CreateTable(err error) bool {
+	switch nerr := errors.Cause(err).(type) {
+	case *terror.Error:
+		if nerr.Code() == tmysql.ErrInvalidDDLJob {
+			return true
+		}
```
Comment (YuJuncen) [inline br/pkg/utils/retry.go:148]:
```suggestion
		return nerr.Code() == tmysql.ErrInvalidDDLJob
```

File: br/tests/br_full_ddl/run.sh [hunk 3]
```diff
@@ -110,6 +111,20 @@ fi
 # clear restore environment
 run_sql "DROP DATABASE $DB;"
 
+# restore full
+echo "restore start..."
+export GO_FAILPOINTS="github.com/pingcap/tidb/br/pkg/restore/restore-createtables-error=return(true)"
+run_br restore full -s "local://$TEST_DIR/$DB" --pd $PD_ADDR --log-file $RESTORE_LOG --ddl-batch-size=128 || { cat $RESTORE_LOG; }
+export GO_FAILPOINTS=""
+
+pause_count=$(cat $RESTORE_LOG | grep "panic"| wc -l | xargs)
```
Comment (YuJuncen) [inline br/tests/br_full_ddl/run.sh:120]: maybe `panic_count` instead of `pause_count`?

(Also, the `xargs` may be removed?)

**Review by YuJuncen — APPROVED** [review]
rest lgtm

File: br/tests/br_full_ddl/run.sh (hunk 3 above)
Comment (fengou1) [reply br/tests/br_full_ddl/run.sh]: agreed.

File: br/pkg/restore/client.go (hunk 1 above)
Comment (fengou1) [reply br/pkg/restore/client.go]: fixed

**Review by kennytm — APPROVED** [review]

**Pull request merged**